### PR TITLE
Fix #74

### PR DESCRIPTION
**Pull request by romulocintra**
closes #74

### CODE REVIEW

**Comment by sffc** [comment]
Hmm. Before, we had

```
        1. Let _value_ be ? Get(_options_, _property_).
        1. If _value_ is *undefined*, return _fallback_.
        1. If _value_ is *true*, return _trueValue_.
        1. Let _valueBoolean_ be ToBoolean(_value_).
        1. If _valueBoolean_ is *false*, return _falsyValue_.
        1. Let _value_ be ? ToString(_value_).
        1. If _values_ does not contain an element equal to _value_, throw a *RangeError* exception.
        1. Return _value_.
```

What if we just change the exception to return `_trueValue_`, like

```
        1. Let _value_ be ? Get(_options_, _property_).
        1. If _value_ is *undefined*, return _fallback_.
        1. If _value_ is *true*, return _trueValue_.
        1. Let _valueBoolean_ be ToBoolean(_value_).
        1. If _valueBoolean_ is *false*, return _falsyValue_.
        1. Let _value_ be ? ToString(_value_).
        1. If _values_ does not contain an element equal to _value_, return _trueValue_.
        1. Return _value_.
```

**Comment by romulocintra** [comment]
> What if we just change the exception to return `_trueValue_`, like
> 
> ```
>         1. Let _value_ be ? Get(_options_, _property_).
>         1. If _value_ is *undefined*, return _fallback_.
>         1. If _value_ is *true*, return _trueValue_.
>         1. Let _valueBoolean_ be ToBoolean(_value_).
>         1. If _valueBoolean_ is *false*, return _falsyValue_.
>         1. Let _value_ be ? ToString(_value_).
>         1. If _values_ does not contain an element equal to _value_, return _trueValue_.
>         1. Return _value_.
> ```

In that case, all strings not contained in list of valid ones will have _trueValue_  behaviour of "always" that might be confusing due to that defaultGrouping it's auto. And how about the use case string "false"?

File: negotiation/diff.emu [hunk 1]
```diff
@@ -333,15 +333,21 @@
...
-        1. If _valueBoolean_ is *false*, return _falsyValue_.
-        1. Let _value_ be ? ToString(_value_).
-        1. If _values_ does not contain an element equal to _value_, throw a *RangeError* exception.
-        1. Return _value_.
+        1. If _value_ is not *undefined*, then
+          1. Assert: _type_ is *"boolean"* or *"string"*.
+          1. If _type_ is *"boolean"*, then
+            1. Let _value_ be ToBoolean(_value_)
+            1. If _value_ is *true*, return _trueValue_.
+            1. If _value_ is *false*, return _falsyValue_.
```
Comment (sffc) [inline negotiation/diff.emu:343]: This section doesn't make sense to me.

- `_type_` is not defined anywhere I can see
- If `_type_` is already a boolean, then why do we need `ToBoolean`?

File: negotiation/diff.emu [hunk 2]
```diff
@@ -333,15 +333,21 @@
...
+          1. If _type_ is *"string"*, then
+            1. Let _value_ be ? ToString(_value_).
+            1. If _value_ contains *true*, return _trueValue_.
+            1. If _value_ contains *false*, return _falsyValue_.
```
Comment (sffc) [inline negotiation/diff.emu:347]: I think `contains` is too broad. It would catch things like "construe" or "falsetto".

**Comment by sffc** [comment]
OK, how about

```
        1. Let _value_ be ? Get(_options_, _property_).
        1. If _value_ is *undefined*, return _fallback_.
        1. If _value_ is *true*, return _trueValue_.
        1. Let _valueBoolean_ be ToBoolean(_value_).
        1. If _valueBoolean_ is *false*, return _falsyValue_.
        1. Let _value_ be ? ToString(_value_).
        1. If _values_ does not contain an element equal to _value_, return _fallback_.
        1. Return _value_.
```

This retains the *current web reality* behavior because `true` is currently just `"auto"`.

**Comment by romulocintra** [comment]
> OK, how about
> 
> ```
>         1. Let _value_ be ? Get(_options_, _property_).
>         1. If _value_ is *undefined*, return _fallback_.
>         1. If _value_ is *true*, return _trueValue_.
>         1. Let _valueBoolean_ be ToBoolean(_value_).
>         1. If _valueBoolean_ is *false*, return _falsyValue_.
>         1. Let _value_ be ? ToString(_value_).
>         1. If _values_ does not contain an element equal to _value_, return _fallback_.
>         1. Return _value_.
> ```
> 
> This retains the _current web reality_ behavior because `true` is currently just `"auto"`.

Updated

> OK, how about
> 
> ```
>         1. Let _value_ be ? Get(_options_, _property_).
>         1. If _value_ is *undefined*, return _fallback_.
>         1. If _value_ is *true*, return _trueValue_.
>         1. Let _valueBoolean_ be ToBoolean(_value_).
>         1. If _valueBoolean_ is *false*, return _falsyValue_.
>         1. Let _value_ be ? ToString(_value_).
>         1. If _values_ does not contain an element equal to _value_, return _fallback_.
>         1. Return _value_.
> ```
> 
> This retains the _current web reality_ behavior because `true` is currently just `"auto"`.

This way there is no risk of breaking current web behaviour and the issue it's solved, on the other hand, throwing an error makes it more consistent with `GetOption` and other methods that use it.

**Comment by sffc** [comment]
The *current* behavior is

```
Let _useGrouping_ be ? StringOrBooleanOption(options, "useGrouping", "boolean", undefined, true).
```

This never fails, because the `undefined` means "accept all values".

**Comment by sffc** [comment]
@anba @FrankYFTang @constellation Take note of this change. `useGrouping` should *not* throw a RangeError any more on malformed arguments.

**Comment by FrankYFTang** [comment]
so.. now it will not only accept {'useGrouping': 'true'} but also {'useGrouping': 'HelloWorld'} ? Isn't that too loose?

**Comment by sffc** [comment]
The issue is that it *already* accepts `{'useGrouping': 'HelloWorld'}`

```
> new Intl.NumberFormat("en", {useGrouping: "HelloWorld"}).resolvedOptions().useGrouping
true
```

So this is a web compat fix.

**Comment by romulocintra** [comment]
> new Intl.NumberFormat("en", {useGrouping: "HelloWorld"}).resolvedOptions().useGrouping
FF:
image

and SAF

image

**Pull request merged**